src/radiation/mocmc.cpp

```diff
@@ -0,0 +1,1100 @@
+
+namespace pf = fluid_prim;
+namespace cf = fluid_cons;
+namespace cr = radmoment_cons;
+namespace pr = radmoment_prim;
+namespace ir = radmoment_internal;
+namespace im = mocmc_internal;
+using namespace singularity;
+using namespace singularity::neutrinos;
+using singularity::RadiationType;
+using vpack_types::FlatIdx;
+
+constexpr int MAX_SPECIES = 3;
+
+// TODO(BRR) add options
+KOKKOS_INLINE_FUNCTION int get_nsamp_per_zone(const int &k, const int &j, const int &i,
+                                              const Geometry::CoordSysMeshBlock &geom,
+                                              const Real &rho, const Real &T,
+                                              const Real &Ye, const Real &J,
+                                              const int &nsamp_per_zone_global) {
+
+  return nsamp_per_zone_global;
+}
+
+template <class T>
+void MOCMCInitSamples(T *rc) {
+
+  auto *pmb = rc->GetParentPointer().get();
+  auto &sc = pmb->swarm_data.Get();
+  auto &swarm = sc->Get("mocmc");
+  // auto swarm = pmb->swarm_data.Get()->Get("mocmc");
+  StateDescriptor *rad = pmb->packages.Get("radiation").get();
+  auto rng_pool = rad->Param<RNGPool>("rng_pool");
+
+  // Meshblock geometry
+  const auto geom = Geometry::GetCoordinateSystem(rc);
+  const IndexRange &ib = pmb->cellbounds.GetBoundsI(IndexDomain::interior);
+  const IndexRange &jb = pmb->cellbounds.GetBoundsJ(IndexDomain::interior);
+  const IndexRange &kb = pmb->cellbounds.GetBoundsK(IndexDomain::interior);
+  const int &nx_i = pmb->cellbounds.ncellsi(IndexDomain::interior);
+  const int &nx_j = pmb->cellbounds.ncellsj(IndexDomain::interior);
+  const int &nx_k = pmb->cellbounds.ncellsk(IndexDomain::interior);
+  const Real &dx_i = pmb->coords.dx1f(pmb->cellbounds.is(IndexDomain::interior));
+  const Real &dx_j = pmb->coords.dx2f(pmb->cellbounds.js(IndexDomain::interior));
+  const Real &dx_k = pmb->coords.dx3f(pmb->cellbounds.ks(IndexDomain::interior));
+  const Real &minx_i = pmb->coords.x1f(ib.s);
+  const Real &minx_j = pmb->coords.x2f(jb.s);
+  const Real &minx_k = pmb->coords.x3f(kb.s);
+
+  // Microphysics
+  auto opac = pmb->packages.Get("opacity");
+  const auto d_opac = opac->template Param<Opacity>("d.opacity");
+  StateDescriptor *eos = pmb->packages.Get("eos").get();
+
+  // auto &dnsamp = rc->Get(mocmc_internal::dnsamp);
+  std::vector<std::string> variables{pr::J,           pr::H,  pf::density, pf::velocity,
+                                     pf::temperature, pf::ye, im::dnsamp};
+  PackIndexMap imap;
+  auto v = rc->PackVariables(variables, imap);
+
+  auto pJ = imap.GetFlatIdx(pr::J);
+  auto pH = imap.GetFlatIdx(pr::H);
+  auto pdens = imap[pf::density].first;
+  auto pv = imap.GetFlatIdx(fluid_prim::velocity);
+  auto pT = imap[pf::temperature].first;
+  auto pye = imap[pf::ye].first;
+  auto dn = imap[im::dnsamp].first;
+
+  const int nblock = v.GetDim(5);
+  PARTHENON_REQUIRE_THROWS(nblock == 1, "Packing not currently supported for swarms");
```

Comment:
You could just make this function only take `MeshBlockData` rather than template it for now.

Reply:
Since we'll eventually go to packing with swarms presumably I'd rather not remove the templating boilerplate now only to reintroduce it later

Reply:
I mean yeah write the code with the packed API, but if you just change the type and remove the template it's like a 2 line change and you catch the error at compile time instead of run time? idk nbd either way